target/linux/ipq806x/base-files/lib/upgrade/buffalo.sh

```diff
@@ -0,0 +1,182 @@
+#!/bin/sh
+# Copyright (C) 2018 OpenWrt.org
+#
+
+. /lib/functions.sh
+
+# The mtd partition 'ubi' and 'rootfs_1' on NAND flash are os-image
+# partitions. These partitions are called as "Bank1/Bank2" in U-Boot
+# on WXR-2533DHP, and they are checked conditions when booting.
+# Then, U-Boot checks kernel and rootfs volumes in ubi, but U-Boot
+# needs "ubi_rootfs" as rootfs volume name. This name cannot use as
+# rootfs volume name in OpenWrt, so we need to prepare dummy rootfs
+# data with squashfs format. And, U-Boot checks the checksum at the
+# end of rootfs (ubi_rootfs).
```

Comment:
maybe you can use the ubi_rootfs dummy as the main rootfs. OpenWrt does not strictly need the rootfs to be named "rootfs" (but if there is a rootfs it will do the automount). it can be anything you want (ubi volume name restriction apply) as long as you pass the kernel the right ubi.block=xyz via the [cmdline/bootargs](https://www.systutorials.com/linux-kernels/474146/ubi-r-o-block-driver-on-top-of-ubi-volumes-linux-3-15/).